TypeCobol.LanguageServer/Processor/DataLayoutProcessor.cs

```diff
@@ -159,57 +77,92 @@ static object GetEnd(long start, long length)
+        {
+            var program = GetProgram(compilationUnit, position);
+
+            return CollectInProgram(program, convert);
+        }
+
+        private static Program GetProgram(CompilationUnit compilationUnit, Position position)
         {
+            Debug.Assert(compilationUnit != null);
+            Debug.Assert(position != null);
             var location = CodeElementLocator.FindCodeElementAt(compilationUnit, position);
             // Get the node corresponding to the position (if null use the main program)
             var locationNode = location.Node ?? (compilationUnit.ProgramClassDocumentSnapshot.Root?.MainProgram);
-            if (locationNode == null)
+            if (locationNode == null || locationNode.GetProgramNode() == null)
             {
                 throw new Exception($"No program found in: '{compilationUnit.TextSourceInfo.Name}'"); ;
             }
 
-            var dataLayoutNodes = new List<Tuple<int, DataDefinition, int>>();
-            var program = locationNode.GetProgramNode();
-            root = program?.Name;
-            DataDivision dataDivision = program?.GetChildren<DataDivision>()?.FirstOrDefault();
+            return locationNode.GetProgramNode();
+        }
+
+        private DataLayoutNode CollectInProgram(Program program, Action<DataLayoutNode> convert)
+        {
+            var rootDLN = DataLayoutNodeBuilder.From(program);
+
+            DataDivision dataDivision = program.GetChildren<DataDivision>()?.FirstOrDefault();
             if (dataDivision != null)
             {
+                var sectionsDLN = new List<DataLayoutNode>();
+
                 // Consider data declared in the Working and Local storage sections
-                CollectDataLayoutNodesInSection(dataDivision.WorkingStorageSection);
-                CollectDataLayoutNodesInSection(dataDivision.LocalStorageSection);
+                CollectInSection(dataDivision.WorkingStorageSection);
+                CollectInSection(dataDivision.LocalStorageSection);
                 // Consider also data declared in the Linkage section
-                CollectDataLayoutNodesInSection(dataDivision.LinkageSection);
-            }
+                CollectInSection(dataDivision.LinkageSection);
 
-            return dataLayoutNodes;
+                rootDLN.children = sectionsDLN.ToArray();
 
-            void CollectDataLayoutNodesInSection(DataSection section)
-            {
-                if (section != null)
+                void CollectInSection(DataSection dataSection)
                 {
-                    CollectDataLayoutNodes(0, section, 0);
+                    if (dataSection != null)
+                    {
+                        var sectionDLN = DataLayoutNodeBuilder.From(dataSection);
+                        sectionsDLN.Add(sectionDLN);
+                        sectionDLN.children = CollectDataLayoutNodes(dataSection, sectionDLN).ToArray();
```

Comment:
This line is weird: the `CollectDataLayoutNodes` method takes the parent as argument but the caller has to set the children by himself ?

To simplify:
- use `List<DataLayoutNode>` instead of array in `DataLayoutNode` for the `children` field
- do not return anything from `CollectDataLayoutNodes`
- add children directly during the recursive build